schedule/scheduler.go

```diff
@@ -14,159 +18,309 @@ import (
+
+			newTask.log = log.With("id", newTask.id, "name", newTask.Name)
+			newTask.log.Infow("new job received")
+
+			go s.execute(newTask, s.workerJobComplete)
+		case <-s.scheduledJobComplete:
+			// A job has completed
+			s.scheduledJobsRunning--
+			if s.scheduledJobsRunning == 0 {
+				log.Info("no scheduled jobs running")
 			}
+		case <-s.workerJobComplete:
+			s.workerJobsRunning--
+			if s.workerJobsRunning == 0 {
+				log.Info("no worker jobs running")
+			}
+		}
+	}
+}
 
-			// Keep this task running forever
-			doneFirstRun := false
-			for {
+func (s *Scheduler) StartJob(id JobID) error {
+	s.jobsMu.Lock()
+	defer s.jobsMu.Unlock()
 
-				// Is the context done?
-				select {
-				case <-ctx.Done():
-					return
-				default:
-				}
+	job, ok := s.jobs[id]
+	if !ok {
+		return xerrors.Errorf("starting worker ID: %d not found", id)
+	}
 
-				if doneFirstRun {
-					log.Infow("restarting task", "task", tc.Name, "delay", tc.RestartDelay)
-					if tc.RestartDelay > 0 {
-						time.Sleep(tc.RestartDelay)
-					}
-				} else {
-					log.Infow("running task", "task", tc.Name)
-					doneFirstRun = true
-				}
+	if job.running {
+		return xerrors.Errorf("starting worker ID: %d already running", id)
+	}
+	job.log.Info("starting job")
+	go s.execute(job, s.workerJobComplete)
+	return nil
+}
+
+func (s *Scheduler) StopJob(id JobID) error {
+	s.jobsMu.Lock()
+	defer s.jobsMu.Unlock()
+
+	job, ok := s.jobs[id]
+	if !ok {
+		return xerrors.Errorf("starting worker ID: %d not found", id)
+	}
+
+	if !job.running {
+		return xerrors.Errorf("starting worker ID: %d already running", id)
+	}
+
+	job.log.Info("stopping job")
+	job.cancel()
+	return nil
+}
+
+type JobResult struct {
+	ID                  JobID
+	Running             bool
+	Name                string
+	RestartOnFailure    bool
+	RestartOnCompletion bool
+	RestartDelay        time.Duration
+}
+
+var InvalidJobID = JobID(0)
+
+type JobID int
+
+func (s *Scheduler) Jobs() []JobResult {
+	s.jobsMu.Lock()
+	defer s.jobsMu.Unlock()
+	if len(s.jobs) == 0 {
+		return nil
+	}
+	var out []JobResult
+	for _, j := range s.jobs {
+		out = append(out, JobResult{
+			ID:                  j.id,
+			Running:             j.running,
+			Name:                j.Name,
+			RestartOnFailure:    j.RestartOnFailure,
+			RestartOnCompletion: j.RestartOnCompletion,
+			RestartDelay:        j.RestartDelay,
+		})
+	}
+	return out
+}
 
-				err := tc.Task.Run(ctx)
-				if err != nil {
-					if errors.Is(err, context.Canceled) {
-						break
-					}
-					log.Errorw("task exited with failure", "task", tc.Name, "error", err.Error())
-
-					if !tc.RestartOnFailure {
-						// Exit the task
-						break
-					}
-				} else {
-					log.Infow("task exited cleanly", "task", tc.Name)
-
-					if !tc.RestartOnCompletion {
-						// Exit the task
-						break
-					}
+func (s *Scheduler) execute(tc *JobConfig, complete chan struct{}) {
+	ctx, cancel := context.WithCancel(s.context)
+	tc.cancel = cancel
+	tc.running = true
```

Comment:
Need to look at the locking here. tc.running is written without a lock but up at line 242 it's also read by a function that could be in a different goroutine. `go test -race` confirms that there is a race here. I also points to tc.cancel being racy probably because you call it in StopJob.

Reply:
Race's fixed in latest commit:
```
┬─[frrist@cypress:~/s/g/f/sentinel-visor]─[10:42:42 AM]─[G:frrist/watch-workers]
╰─>$ git rev-parse HEAD
0f83f196a152788501d6516f681c98d7a6629728

┬─[frrist@cypress:~/s/g/f/sentinel-visor]─[10:42:44 AM]─[G:frrist/watch-workers]
╰─>$ go test -count=1 -race ./schedule/
ok      github.com/filecoin-project/sentinel-visor/schedule     0.042s
```